cabal-install/Distribution/Solver/Modular/Solver.hs

```diff
@@ -138,6 +136,19 @@ solve sc cinfo idx pkgConfigDB userPrefs userConstraints userGoals =
                      $ addLinking
                      $ buildTree idx (independentGoals sc) userGoals
 
+    -- Counting conflicts and reordering goals interferes, as both are strategies to
+    -- change the order of goals. When count-conflicts is set, we therefore interpret
+    -- reorder-goals to only prefer goals with 0 or 1 enabled choice.
+    --
+    -- In the past, we used P.firstGoal to trim down the goal choice nodes to just a
+    -- single option. This was a way to work around a space leak that was unnecessary
+    -- and is now fixed, so we no longer do it.
+    --
+    goalChoiceHeuristics
+      | asBool (reorderGoals sc) && asBool (countConflicts sc) = P.preferReallyEasyGoalChoices
+      | asBool (reorderGoals sc)                               = P.preferEasyGoalChoices
+      | otherwise                                              = id {- P.firstGoal -}
```

Comment:
I think these comments would be clearer if they described which goals are pruned in each case.